<a href="https://colab.research.google.com/github/willyan041107/DS1002/blob/main/Assignments/14_TextGeneration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Final homework assignment:  generating text from a corpus

Due Friday 3 May at 5 pm

We begin by loading all of Shakespeare's plays into a text file, and separating them into individual words.  This is made easy by the fact that someone already made a giant text file of these plays for us, and put it online.

[Note that by changing the url in the following code chunk, you can load any text file that lives online in this way . . . see below.]

Run the following code.  It should output a list of ten words, starting with 'From' and ending with 'rose'.

In [2]:
from urllib.request import urlopen            #package for reading info from the web
url='https://ocw.mit.edu/ans7870/6/6.006/s08/lecturenotes/files/t8.shakespeare.txt'   #web address
text = urlopen(url).read().decode('utf-8')    #open web address, read file, convert to text
shakespeare_words=text.split()[1651:]         #split text into words (not including preamble)
shakespeare_words[:10]                        #display the first 10 words

['From',
 'fairest',
 'creatures',
 'we',
 'desire',
 'increase,',
 'That',
 'thereby',
 "beauty's",
 'rose']

## 1. `nextwords` practice

Create a practice function called `createNextwordsPractice` which takes a list of words called `wordlist` and returns a dictionary called `nextwordsPractice`. The keys of `nextwordsPractice` should be all the words in `wordlist`, and the value for any given key should be a list of all the words that follow the key word somewhere in the original `wordlist`.



In [15]:
def createNextwordsPractice(wordlist):
    nextwordsPractice = {}
    for i in range(len(wordlist) - 1):
        word = wordlist[i]
        next_word = wordlist[i + 1]
        if word not in nextwordsPractice:
            nextwordsPractice[word] = [next_word]
        else:
            nextwordsPractice[word].append(next_word)
    return nextwordsPractice

Run the following code:

In [16]:
practice = createNextwordsPractice(shakespeare_words)
practice['From']

['fairest',
 'his',
 'sullen',
 'limits',
 'hands',
 'whence',
 'where',
 'me',
 'this',
 'thy',
 'hence',
 'thee,',
 'you',
 'heaven',
 'her',
 'simple',
 'whence',
 'below',
 'lowest',
 'courtly',
 'son',
 'the',
 'the',
 'Sicyon,',
 'Alexandria',
 'Silvius,',
 'mine',
 'the',
 'firm',
 'Egypt',
 'Antony',
 'thine',
 'that',
 'my',
 'me',
 'henceforth',
 'tyrant',
 'seventeen',
 'miserable',
 'whom',
 'my',
 'whence,',
 'me',
 'me',
 'slaves',
 'every',
 'rascals',
 'whom',
 'where',
 "th'",
 'Rome',
 'off',
 'these',
 'twelve',
 "th'",
 'whence?',
 'Rome.',
 'thy',
 'Gallia',
 'fairies',
 'off',
 "th'",
 'most',
 'thy',
 'youth',
 'every',
 'whose',
 'so',
 'one',
 'this',
 'the',
 'east',
 'action',
 'my',
 'this',
 'Leonati',
 'her',
 'chance',
 'south',
 'our',
 'whence',
 'the',
 'top',
 'this',
 'our',
 'that',
 'me,',
 'fashion',
 'the',
 'another',
 'Hamlet?',
 'whence',
 'whom',
 "Rumour's",
 'whence',
 'the',
 'a',
 'a',
 'enemies',
 'every',
 'this',
 'the',
 'open',
 'who

In [17]:
practice['rose']

['might',
 'looks',
 'is',
 'of',
 'Of',
 'may',
 'will',
 'of',
 'From',
 'and',
 'of',
 'both',
 'with',
 'from',
 'with',
 'with',
 'side.',
 'red,',
 'too.',
 'in',
 'a',
 'a',
 'I',
 'against',
 'that',
 'and',
 'from',
 'again,',
 'Lest',
 'against',
 'Than',
 "distill'd",
 'on',
 'up',
 'in',
 'wither.',
 'and',
 'By',
 'again?',
 'he']

In [18]:
practice['rose.']

['It']

## 2. `nextwords` for real:  new sentences.

Next, write a new function called `createNextwords` which takes a `wordlist` and returns a dictionary of `nextwords`, but now with a new additional feature.  
In addition to the keys from problem 1, add a special key to the dictionary, which will be the symbol '%'.  This key will signify the first word of a sentence. When scanning the wordlist to construct the `nextwords` dictionary, normally you would add `wordlist[i+1]` to the list `nextwords[wordlist[i]]`. However, if `wordlist[i]` ends with a period, then you should instead add `wordlist[i+1]` to the list nextwords['%']. Also add the very first word of `wordlist` to nextwords['%'].


You should be able to produce this output:
```
> shakespeare_nextwords=createNextwords(shakespeare_words)
> shakespeare_nextwords['Rose']
['at', 'or', 'or', 'and']

> shakespeare_nextwords['rose.']
[]

> len(shakespeare_nextwords['%'])
76543
```

In [50]:
def createNextwords(wordlist):
    nextwords = {'%': [wordlist[0]]}
    for i in range(len(wordlist) - 1):
        current = wordlist[i]
        next_word = wordlist[i + 1]
        if current not in nextwords:
            nextwords[current] = []
        nextwords[current].append(next_word)

        if current[-1] == '.':
            nextwords[current] = []
            nextwords['%'].append(next_word)
    return nextwords

**Run** the following code:

In [51]:
shakespeare_nextwords = createNextwords(shakespeare_words)

print(shakespeare_nextwords['Rose'])
print(len(shakespeare_nextwords['%']))

['at', 'or', 'or', 'and']
76543


In [52]:
print(len(shakespeare_nextwords['rose']))

40


In [53]:
print(shakespeare_nextwords['rose.'])

[]


## 3.  Make your own Shakespeare play

Okay, that may be a bit of an exaggeration.  Anyway, next up is text generation.  
We'll need the choice function:

In [54]:
from random import choice

This function chooses a random element from a list.  Run this code a few times:

In [55]:
choice([1, 17, 34, 289])

289

Now, create a function called `makePlay(nextwords,n)` which takes a dictionary `nextwords` and a number `n`, and follows these steps:
1. Define an empty `newplay` string.  (I did this for you.)
2. Set `currentword` equal to '%'.
3. Start a for-loop that will run `n` times.
4. Randomly choose a word from the list `nextwords[currentword]`.
5. Add the chosen word to `newplay`, followed by a space `' '` character.
6. If the chosen word ends in a period, add a newline character '\n' and set `currentword` equal to `%`.
7. Otherwise, set `currentword` equal to the chosen word.
8. Go back to step 4.
9. When all `n` iterations have finished, your function should return `newplay`.

In [56]:
def makePlay(nextwords, n):
    newplay = ''
    currentword = '%'
    for _ in range(n):
        currentword = choice(nextwords[currentword])
        newplay += currentword + ' '
        if currentword[-1] == '.':
            newplay += '\n'
            currentword = '%'
    return newplay

Now create a 1000 word Shakesperian play by running the following code:

In [57]:
nextwords=createNextwords(shakespeare_words)
newplay=makePlay(nextwords,1000)
print(newplay)

DAVY. 
O Seigneur Dieu! les oeuvres. 
May be too bad, going to GLOUCESTER] Come hither, come by. 
Most certain. 
What treason for my lord, My father's right. 
Exeunt SCENE I. 
My father is set! Our wooing mind to notify unto thy friends, and confusion! All this satisfaction canst not wounded? [To the infirmity Will You worthy comfort, And bless thee, make a stick boarish fangs. 
This shall find means To want to funerals; The urging of Gonzago'? 1. 
What's your pleasure, sweet soul, I'll run, fiend; obey you, and die no boot. 
He shall laugh mortal. 
I read who did hear, And top of you! BEROWNE. 
Hath he doth say I am subject to mine, And I doubt Will you were that the name her in lions' stead: Sheep run at, weapons round hose, a lady That left a time of Norfolk, rescue, You, Titus Lartius, a sand-bag fastened on such things past, For such man that I am fresh assault. 
If thou not? It is so sick men by day, By certain French; Use us by the kingdom. 
The mouse of our jewels trifles as st

At this point you might enjoy finding other sources of text on the internet and replacing the url at the top of this notebook. You can use song lyrics, news articles, etc. You just need to find files ending with '.txt'. Alternatively, you can copy and paste any block of text from any file into colab, assign it to the variable `text`, and then use wordlist=text.split() to get a list of words.

## Bonus:

Recall the following recursive function that we used to define the Fibonacci numbers:

In [37]:
def fibonacci(n):
  # Base cases: F(0) = 0, F(1) = 1
  if n == 0:
    return 0
  elif n == 1:
    return 1
  else:
    # Recursive case: F(n) = F(n-1) + F(n-2)
    return fibonacci(n - 1) + fibonacci(n - 2)


Here is the question.  If you run this line of code:  
`fibonacci(17)`  
how many times does the function `fibonacci()` get called?  
(The answer is not 1597, which is the value returned by the function.)

In [40]:
def fibonacci_count(n, count=0):
    count += 1
    if n == 0:
        return 0, count
    elif n == 1:
        return 1, count
    else:
        fib_n_1, count = fibonacci_count(n - 1, count)
        fib_n_2, count = fibonacci_count(n - 2, count)
        return fib_n_1 + fib_n_2, count

In [41]:
fibonacci_count(17)

(1597, 5167)

Therefore, the fibonacci() function is called 5167 times when computing fibonacci(17).